relational_operators/SelectOperator.cpp

```diff
@@ -54,43 +54,93 @@ bool SelectOperator::getAllWorkOrders(
+                                  predicate,
+                                  simple_projection_,
+                                  simple_selection_,
+                                  selection,
+                                  output_destination,
+                                  storage_manager),
+              op_index_);
+        }
       }
       started_ = true;
     }
     return started_;
   } else {
-    while (num_workorders_generated_ < input_relation_block_ids_.size()) {
-      container->addNormalWorkOrder(
-          new SelectWorkOrder(
-              input_relation_,
-              input_relation_block_ids_[num_workorders_generated_],
-              predicate,
-              simple_projection_,
-              simple_selection_,
-              selection,
-              output_destination,
-              storage_manager),
-          op_index_);
-      ++num_workorders_generated_;
+    if (input_relation_.hasPartitionScheme()) {
+      const PartitionScheme &part_scheme =
+          input_relation_.getPartitionScheme();
+      int num_partitions = part_scheme.getPartitionSchemeHeader().getNumPartitions();
+      for (int part_id = 0; part_id < num_partitions; ++part_id) {
+        while (num_workorders_generated_in_partition_[part_id] <
+               input_relation_block_ids_in_partition_[part_id].size()) {
+          container->addNormalWorkOrder(
+              new SelectWorkOrder(
+                  input_relation_,
+                  input_relation_block_ids_in_partition_
+                      [part_id]
+                      [num_workorders_generated_in_partition_[part_id]],
+                  predicate,
+                  simple_projection_,
+                  simple_selection_,
+                  selection,
+                  output_destination,
+                  storage_manager,
+                  placement_scheme_.getNUMANodeForBlock(
+                    input_relation_block_ids_in_partition_[part_id][num_workorders_generated_in_partition_[part_id]])),
+              op_index_);
+          ++num_workorders_generated_in_partition_[part_id];
+        }
+      }
+    } else {
+      while (num_workorders_generated_ < input_relation_block_ids_.size()) {
+        container->addNormalWorkOrder(
+            new SelectWorkOrder(
+                input_relation_,
+                input_relation_block_ids_[num_workorders_generated_],
+                predicate,
+                simple_projection_,
+                simple_selection_,
+                selection,
+                output_destination,
+                storage_manager),
+            op_index_);
+        ++num_workorders_generated_;
+      }
     }
     return done_feeding_input_relation_;
   }
 }
 
 void SelectWorkOrder::execute() {
   BlockReference block(
-      storage_manager_->getBlock(input_block_id_, input_relation_));
+      storage_manager_->getBlock(input_block_id_, input_relation_, getPreferredNUMANodes()[0]));
```

Comment:
This may segfault if getPreferredNUMANodes() returns an empty vector. Error handling is needed to avoid such a case. Also, adding a note on why we pick only the first NUMA node from the vector should be helpful.